Automattic/jetmon
# Fix incorrect SPS metric calculations and flush cached child stats before recycling

This PR addresses some metric calculation issues with how SPS metrics are calculated and fixes an issue where Jetmon child processes can recycle before flushing their stats.

The SPS (sites per second) calculation did not account for the stats flush interval set by the `STATS_UPDATE_INTERVAL_MS` config. This meant that, for a configured flush every 5 seconds, the reported SPS value was inflated 5x. This change fixes the calculation.

Jetmon child processes could recycle with cached stats. All the calls to `process.exit` are now preceded with a call to the new `HttpChecker.sendStats` function which replaces the anonymous function used to set up the stats flush interval that runs every `STATS_UPDATE_INTERVAL_MS` MS. This change should result in much more accurate site processing stats since children will no longer recycle with unsent stats.

While making this change, I also noticed that a site success count was incorrectly reported as a total count. This metric is now changed to a success count. A new total count metric is added to provide a correct total count.

Itemized changes to StatsD metrics:

- `stats.sites.sps.count` now represents an average per second over the flush interval.
- `stats.sites.total.count` now represents the total number of sites processed in the flush interval rather than the success count.
- `stats.sites.success.count` was added to represent the count of sites that are up.

## Testing Instructions

1. Check out this PR's branch `fix/sps-metric-calculation`.
2. Ensure that the test environment and config is set up in a way to ensure that workers will attempt to recycle in a reasonable amount of time. Here are the key details from my testing environment:
    - Set the following in `config/config.json`:
      - `NUM_WORKERS` to `800`
      - `NUM_TO_PROCESS` to `20`
      - `DATASET_SIZE` to `50`
      - `WORKER_MAX_CHECKS` to `200`
      - `BUCKET_NO_MIN` to `0`
      - `BUCKET_NO_MAX` to `511`
      - `STATS_UPDATE_INTERVAL_MS` to `10000`
      - `USE_VARIABLE_CHECK_INTERVALS` to `true`
    - Filled my test database with 10,000 test sites to ensure that child processes recycle frequently just due to the number of sites processed.
3. Run Jetmon.
4. Load up Graphite ([https://localhost::8088](https://localhost::8088)) and ensure that the following metrics appear:
    - `stats.com.jetpack.jetmon.jetmon.docker.stats.sites.sps.count`
    - `stats.com.jetpack.jetmon.jetmon.docker.stats.sites.total.count`
    - `stats.com.jetpack.jetmon.jetmon.docker.stats.sites.success.count`
5. If you want to confirm that stats that would be lost are now sent, you can add the following code just above line 203 in `lib/httpcheck.js`.
```console.log( 'Flushing stats due to max checks', workerTotals );```
After restarting Jetmon and monitoring `docker compose logs -f` output, you should quickly see output such as the following:
```Flushing stats due to max checks { '0': 0, '1': 6, '2': 1 }```
Note the non-zero values which indicate stats that would be lost before this change.
